go/osagent/osagent.go

```diff
@@ -645,3 +745,609 @@ func AbortSeed(seedId string) error {
+		}
+		if config.Config.CompressLogicalBackup {
+			cmd += fmt.Sprintf(" | gzip > %s", path.Join(BackupFolder, mysqlbackupCompressedFileName))
+		} else {
+			cmd += fmt.Sprintf(" > %s", path.Join(BackupFolder, mysqlbackupFileName))
+		}
+	}
+	err = commandRun(
+		fmt.Sprintf(cmd),
+		func(cmd *exec.Cmd) {
+			activeCommands[seedId] = cmd
+			log.Debug("StartLocalBackup command completed")
+		})
+	if err != nil {
+		return "", log.Errore(err)
+	}
+	return BackupFolder, err
+}
+
+func backupMySQLUsers(seedId string) (err error) {
+	if len(config.Config.MySQLBackupUsersOnTargetHost) > 0 {
+		statement, err := dbagent.GenerateBackupForUsers(config.Config.MySQLBackupUsersOnTargetHost)
+		if err != nil {
+			return log.Errore(err)
+		}
+		f, err := os.OpenFile(path.Join(config.Config.MySQLBackupDir, mysqlUserBackupFileName), os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0666)
+		defer f.Close()
+		if err != nil {
+			return log.Errore(err)
+		}
+		_, err = f.WriteString(statement)
+		if err != nil {
+			return log.Errore(err)
+		}
+	} else {
+		cmd := fmt.Sprintf("mysqldump --master-data=2 --set-gtid-purged=OFF --user=%s --password=%s --port=%d --single-transaction mysql %s > %s",
+			config.Config.MySQLTopologyUser, config.Config.MySQLTopologyPassword, config.Config.MySQLPort, strings.Join(mysqlUsersTables[:], " "), path.Join(config.Config.MySQLBackupDir, mysqlUserBackupFileName))
+		err = commandRun(
+			fmt.Sprintf(cmd),
+			func(cmd *exec.Cmd) {
+				activeCommands[seedId] = cmd
+				log.Debug("Backing up MySQL users")
+			})
+	}
+	cmd := fmt.Sprintf("echo 'FLUSH PRIVILEGES;' >> %s", path.Join(config.Config.MySQLBackupDir, mysqlUserBackupFileName))
+	err = commandRun(
+		fmt.Sprintf(cmd),
+		func(cmd *exec.Cmd) {
+			activeCommands[seedId] = cmd
+			log.Debug("Adding FLUSH PRIVILEGES command to MySQL users backup")
+		})
+	return err
+}
+
+func restoreMySQLUsers(seedId string, backupFolder string) error {
+	cmd := fmt.Sprintf("mysql -u%s -p%s --port %d mysql < %s", config.Config.MySQLTopologyUser, config.Config.MySQLTopologyPassword, config.Config.MySQLPort, path.Join(config.Config.MySQLBackupDir, mysqlUserBackupFileName))
+	err := commandRun(
+		fmt.Sprintf(cmd),
+		func(cmd *exec.Cmd) {
+			activeCommands[seedId] = cmd
+			log.Debug("Restoring MySQL users")
+		})
+	return log.Errore(err)
+}
+
+func ReceiveBackup(seedId string, seedMethod string, backupFolder string) error {
+	var cmd string
+	var err error
+	var MySQLInnoDBLogDir string
+	if !contains(seedMethod, seedMethods) {
+		return log.Errorf("Unsupported seed method")
+	}
+	if config.Config.MySQLBackupOldDatadir {
+		if err := MySQLStop(); err != nil {
+			return log.Errore(err)
+		}
+		cmd := fmt.Sprintf("tar zcfp %s -C %s .", path.Join(config.Config.MySQLBackupDir, mysqlBackupDatadirName), config.Config.MySQLDataDir)
+		err := commandRun(
+			fmt.Sprintf(cmd),
```

Comment:
same as above

Reply:
Taking a break. Review still not complete. TODO Marking this for next iteration.

Reply:
fixed in 3ce673a